main.go

```diff
@@ -0,0 +1,180 @@
+	cmdIptables.Stderr = os.Stderr
+
+	if err := cmdIptables.Run(); err != nil {
+		code := 1
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			code = exitErr.ExitCode()
+		} else {
+			// If it's not an ExitError, the command probably didn't finish and something
+			// else failed, which means it might not had outputted anything. In that case,
+			// print the error message just in case.
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+		}
+		os.Exit(code)
+	}
+}
+
+// iptablesInstallation represents the set of iptables-*-save binaries installed in a machine.
+// It is expected the machine supports both nft and legacy modes. This can be implemented by
+// calling directly iptables-*-save, xtables, etc. The implementation should accept the same
+// command arguments as the mentioned binaries.
+type iptablesInstallation interface {
+	// legacySave runs a iptables-legacy-save command
+	legacySave(ctx context.Context, out *bytes.Buffer, args ...string) error
+	// legacySaveIP6 runs a ip6tables-legacy-save command
+	legacySaveIP6(ctx context.Context, out *bytes.Buffer, args ...string) error
+	// nftSave runs a iptables-nft-save command
+	nftSave(ctx context.Context, out *bytes.Buffer, args ...string) error
+	// nftSaveIP6 runs a ip6tables-nft-save command
+	nftSaveIP6(ctx context.Context, out *bytes.Buffer, args ...string) error
+}
+
+// detectIptablesMode inspects the current iptables entries and tries to
+// guess which iptables mode is being used: legacy or nft
+func detectIptablesMode(ctx context.Context, iptables iptablesInstallation) iptablesMode {
+	// This method ignores all errors, this is on purpose. We execute all commands
+	// and try to detect patterns in a best effort basis. If somthing fails,
+	// continue with the next step. Worse case scenario if everything fails,
+	// default to nft.
+
+	// In kubernetes 1.17 and later, kubelet will have created at least
+	// one chain in the "mangle" table (either "KUBE-IPTABLES-HINT" or
+	// "KUBE-KUBELET-CANARY"), so check that first, against
+	// iptables-nft, because we can check that more efficiently and
+	// it's more common these days.
+	allMangledNFTRules := &bytes.Buffer{}
+	_ = iptables.nftSave(ctx, allMangledNFTRules, "-t", "mangle")
+	_ = iptables.nftSaveIP6(ctx, allMangledNFTRules, "-t", "mangle")
+	if hasKubeletChains(allMangledNFTRules.Bytes()) {
+		return nft
+	}
+
+	// Check for kubernetes 1.17-or-later with iptables-legacy. We
+	// can't pass "-t mangle" to iptables-legacy-save because it would
+	// cause the kernel to create that table if it didn't already
+	// exist, which we don't want. So we have to grab all the rules.
+	allLegacyRulesOut := &bytes.Buffer{}
+	_ = iptables.legacySave(ctx, allLegacyRulesOut)
+	_ = iptables.legacySaveIP6(ctx, allLegacyRulesOut)
+	allLegacyRules := allLegacyRulesOut.Bytes()
+	if hasKubeletChains(allLegacyRules) {
+		return legacy
+	}
+
+	// With older kubernetes releases there may not be any _specific_
```

Comment:
I think it's acceptable to say that this version of iptables-wrappers only works with kubernetes 1.17 and later, and if you need to support older releases, then use an older version of iptables-wrappers. And then you can remove this code

Reply:
Cool, I just tried to make this as close as possible to the original script. Happy to simplify it if we are ok with dropping support for <1.17